File: cses/2165.py
res = []
 
def solve(k, x, y, z):
    if k == 1:
        res.append((x, y))
    else:
        solve(k - 1, x, z, y)
        solve(1, x, y, z)
        solve(k - 1, z, y, x)
 
k = int(input())
solve(k, 1, 3, 2)
print(len(res))
for a, b in res:
    print(a, b)

File: cses/2165.cpp
#include <bits/stdc++.h>
using namespace std;
 
int n;
vector<pair<int, int>> res;
 
void moveDisk(int number, int from, int to, int intermediary) {
    if(number == 1) {
        res.push_back({from, to});
        return;
    }
    moveDisk(number - 1, from, intermediary, to);
    moveDisk(1, from, to, intermediary);
    moveDisk(number - 1, intermediary, to, from);
}
 
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
 
    cin >> n;
    
    moveDisk(n, 1, 3, 2);
    cout << res.size() << "\n";
    for(auto [x, y] : res) 
        cout << x << " " << y << "\n";
    
    return 0;
}